### PR TITLE
Grid transforms

### PR DESCRIPTION
**Pull request by mibaumgartner**
## Short Description
Introduce Grid Transformations including:
 - [ ] GridTransform Class (generalisation of affine transforms)
 - [ ] Grid Crop
 - [ ] Elastic Deformation
 - [ ] Random Deformation
 - [ ] Radial Deformation

This should be merged after https://github.com/PhoenixDL/rising/pull/31

### PR Implementer
This is a small checklist for the implementation details of this PR.
If you submit a PR, please look at these points (don't worry about the `RisingTeam`
and `Reviewer` workflows, the only purpose of those is to have a compact view of
the steps)

If there are any questions regarding code style or other conventions check out our 
[summary](https://github.com/PhoenixDL/rising/blob/master/CONTRIBUTING.md).

- [ ] Implementation
- [ ] Docstrings & Typing
- [ ] Check `__all__` sections and `__init__`
- [ ] Unittests (look at the line coverage of your tests, the goal is 100%!)
- [ ] Update notebooks & documentation if necessary
- [ ] Pass all tests
- [ ] Add the checksum of the last implementation commit to the Changelog

### RisingTeam
<details>
  <summary>RisingTeam workflow</summary>

  - [ ] Add pull request to project (optionally delete corresponding project note)
  - [ ] Assign correct label (if you don't have permission to do this, someone will do it for you. 
      Please make sure to communicate the current status of the pr.)
  - [ ] Does this PR close an Issue? (add `closes #IssueNumber` at the bottom if 
        not already in description)
</details>

### Reviewer
<details>
  <summary>Reviewer workflow</summary>
  
  - [ ] Do all tests pass? (Unittests, NotebookTests, Documentation)
  - [ ] Does the implementation follow `rising` design conventions?
  - [ ] Are the tests useful? (!!!) Are additional tests needed? 
        Can you think of critical points which should be covered in an additional test?
  - [ ] Optional: Check coverage locally / Check tests locally if GPU is necessary to execute
</details>

### CODE REVIEW

File: rising/transforms/grid.py [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+
+            data[key] = torch.nn.functional.grid_sample(
+                data[key], _grid, mode=self.interpolation_mode,
+                padding_mode=self.padding_mode, align_corners=self.align_corners,
+                **self.kwargs)
+        self.grid = None
+        return data
+
+    @abstractmethod
+    def augment_grid(self, grid: Dict[Tuple, Tensor]) -> Dict[Tuple, Tensor]:
```
Comment (justusschock) [inline rising/transforms/grid.py]: What is this function for?
Maybe it would be better to have this return the grid per default (as done in the affines).

File: rising/transforms/grid.py [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+
+        grid = {}
+        for size in input_size:
+            if tuple(size) not in grid:
+                grid[tuple(size)] = torch.nn.functional.affine_grid(
+                    matrix, size=input_size, align_corners=self.align_corners)
+        return grid
+
+    def __add__(self, other):
+        if not isinstance(other, GridTransform):
```
Comment (justusschock) [inline rising/transforms/grid.py]: Can't we also add the case, where there is just the grid provided and not wrapped with an transform?

File: rising/transforms/grid.py [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+from torch import Tensor
+
+from rising.transforms import AbstractTransform
+from rising.utils.affine import get_batched_eye, matrix_to_homogeneous
+
+
+__all__ = ["GridTransform", "StackedGridTransform"]
+
+
+class GridTransform(AbstractTransform):
```
Comment (justusschock) [inline rising/transforms/grid.py]: What if I somehow get my grid from somewhere else and just want to use it? can we have an optional argument in here?

File: rising/transforms/grid.py [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+                transforms = transforms[0]
+        self.transforms = transforms
+
+    def create_grid(self, input_size: Sequence[Sequence[int]], matrix: Tensor = None) -> \
+            Dict[Tuple, Tensor]:
+        return self.transforms[0].create_grid(input_size=input_size, matrix=matrix)
+
+    def augment_grid(self, grid: Tensor) -> Tensor:
+        for transform in self.transforms:
+            grid = transform.augment_grid(grid)
```
Comment (justusschock) [inline rising/transforms/grid.py]: I'm not sure this will work, if it is not implemented for each of the base transforms...

File: rising/transforms/grid.py [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+
+
+class StackedGridTransform(GridTransform):
+    def __init__(self, *transforms: Union[GridTransform, Sequence[GridTransform]]):
+        super().__init__(keys=None, interpolation_mode=None, padding_mode=None,
+                         align_corners=None)
+        if isinstance(transforms, (tuple, list)):
+            if isinstance(transforms[0], (tuple, list)):
+                transforms = transforms[0]
+        self.transforms = transforms
```
Comment (justusschock) [inline rising/transforms/grid.py]: Can we also include a type check here which checks, that each item is a valid `GridTransform`?

File: rising/transforms/grid.py (hunk 2 above)
Comment (mibaumgartner) [reply rising/transforms/grid.py]: i can add that

File: rising/transforms/grid.py (hunk 3 above)
Comment (mibaumgartner) [reply rising/transforms/grid.py]: same as above?

File: rising/transforms/grid.py (hunk 4 above)
Comment (mibaumgartner) [reply rising/transforms/grid.py]: this is now implemented for all base transforms

File: rising/transforms/grid.py (hunk 1 above)
Comment (mibaumgartner) [reply rising/transforms/grid.py]: This function gets a grid and applies its augmentation to it. It returns a grid for every spatial size which was found inside the keys

File: rising/transforms/grid.py (hunk 3 above)
Comment (justusschock) [reply rising/transforms/grid.py]: I don't get it

Comment (mibaumgartner) [reply rising/transforms/grid.py]: Sorry, thought the upper comment was at the same position. ^^
Yeah, we can add that. I think I would prefer an extra Transform for that, just like the Affine Transform has
